File: 242 - Valid Anagram/solution.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 * Time Complexity - O(n log n)
 * Space Complexity - O(n)
 */
var isAnagram = function(s, t) {
    
    // Using Sorting
    if(s.length != t.length){
        return false;
    }
    
    return s.split('').sort().join('') === t.split('').sort().join('') ? true : false;  
    
};

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 * Time Complexity - O(n)
 * Space Complexity - O(n)
 */
var isAnagram = function(s, t) {

    if(s.length !== t.length){
        return false;
    }

    let arr = Array(26).fill(0);

    for(let i=0; i<s.length; i++){
        arr[s.charCodeAt(i)-97]++;
        arr[t.charCodeAt(i)-97]--;
    }

    for(let val of arr){
        if(val !== 0){
            return false;
        }
    }

    return true;  
    
};

File: 242 - Valid Anagram/solution.java
/*
 * Time Complexity - O(n)
 * Space Complexity - O(n)
 */

public class solution {
    public boolean isAnagram(String s, String t) {
        int[] arr = new int[26];

        for(char ch: s.toCharArray()){
            arr[ch - 'a']++;
        }
        
        for(char ch: t.toCharArray()){
            arr[ch - 'a']--;
        }

        for(int val: arr){
            if(val != 0){
                return false;
            }
        }

        return true;
    }
}
